modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/IgniteH2Indexing.java

```diff
@@ -1294,6 +1307,18 @@ else if (parseRes.isDml()) {
         }
     }
 
+    /** Initiate new cancel holder, chaining cancelation activity with previous state. */
+    private GridQueryCancel wrapCancel(GridQueryCancel inst) {
+        GridQueryCancel cancel0 = new GridMultiStatementQueryCancel();
+        try {
+            inst.add(cancel0::cancel);
```

Comment:
cyclic cancel? it looks incorrect

Reply:
yes cyclic, but i can`t see the problem here, can you clarify ? It just fast return.
```
    public synchronized void cancel() {
        if (canceled)
            return;
```